walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,217 @@
+    for (String attendee : request.getAttendees()) {
+      meetingTimesToAdd = determineAvailableTime(
+          request, firstFlag, meetingTimesToAdd, schedules, attendee, meetingTimes);
+      for (TimeRange adding : meetingTimesToAdd) {
+        meetingTimes.add(adding);
+      }
+      meetingTimesToAdd.clear();
+      firstFlag = 0;
+    }
+
+    // Takes into consideration all optional attenddes to find free slots
+    ArrayList<TimeRange> optionalMeetingTimesToAdd = new ArrayList<TimeRange>();
+    Map<String, ArrayList<TimeRange>> optionalFrees = new HashMap<String, ArrayList<TimeRange>>();
+    for (String optionalAttendee : request.getOptionalAttendees()) {
+      firstFlag = 0;
+      ArrayList<TimeRange> optionalMeetingTimes = (ArrayList<TimeRange>) meetingTimes.clone();
+      optionalMeetingTimesToAdd = determineAvailableTime(request, firstFlag,
+          optionalMeetingTimesToAdd, schedules, optionalAttendee, optionalMeetingTimes);
+      ArrayList<TimeRange> toPut = (ArrayList<TimeRange>) optionalMeetingTimesToAdd.clone();
+      if (!toPut.isEmpty()) {
+        optionalFrees.put(optionalAttendee, toPut);
+      }
+      optionalMeetingTimesToAdd.clear();
+    }
+
+    // If only one optional can attend, return that slot
+    if (optionalFrees.size() == 1) {
+      return optionalFrees.values().iterator().next();
+    }
+
+    // Find slots where the most optionals can attend
+    Map<TimeRange, Integer> freeCounter = new HashMap<TimeRange, Integer>();
+    freeCounter = determineFrequency(freeCounter, optionalFrees, request);
+    int maximum = 0;
+    ArrayList<TimeRange> finalTimes = new ArrayList<TimeRange>();
+    for (TimeRange window : freeCounter.keySet()) {
+      if (freeCounter.get(window) > maximum) {
+        maximum = freeCounter.get(window);
+      }
+    }
+
+    // If no free slots are available, return the mandatory free slots
+    if (maximum == 0) {
+      return meetingTimes;
+    }
+
+    // Otherwise, find the most available windows
+    for (TimeRange window : freeCounter.keySet()) {
+      if (freeCounter.get(window) == maximum) {
+        finalTimes.add(window);
+      }
+    }
+    Collections.sort(finalTimes, TimeRange.ORDER_BY_START);
+    return finalTimes;
+  }
+
+  private Map<String, ArrayList<TimeRange>> initSchedules(
+      MeetingRequest request, Map<String, ArrayList<TimeRange>> schedules) {
+    // Initialize the schedules map, making each attendee completely free
+    for (String requestAttendee : request.getAttendees()) {
+      schedules.put(requestAttendee, new ArrayList(Arrays.asList(TimeRange.WHOLE_DAY)));
+    }
+    for (String optionalAttendee : request.getOptionalAttendees()) {
+      schedules.put(optionalAttendee, new ArrayList(Arrays.asList(TimeRange.WHOLE_DAY)));
+    }
+    return schedules;
+  }
+
+  private Map<String, ArrayList<TimeRange>> scheduleEvent(Event event, String attendee,
+      ListIterator scheduleIterator, Map<String, ArrayList<TimeRange>> schedules) {
```

Comment:
There is no need to pass scheduleIterator, since you already pass schedules as a parameter. Just iterator the schedules in this function.